client/src/components/_common/Button/Button.test.js

```diff
@@ -1,20 +1,160 @@
+  const root = getByRole('button');
+  const text = getByTestId('text');
+  const typeClassName = BTN.TYPE_MAP[type];
+  const sizeClassName = BTN.SIZE_MAP[size];
+  expect(root.className).toMatch('root');
+  expect(root.className).toMatch(new RegExp(typeClassName));
+  expect(root.className).toMatch(new RegExp(sizeClassName));
+  expect(text.className).toMatch('text');
+}
+
+function muteTypeNotLinkNoSizeLog(type, size) {
+  if (type !== 'link' && !size) console.debug = jest.fn();
+}
+
+function isPropertyLimitation(type, size) {
+  let isLimited = false;
+
+  if (
+    type === 'primary' && size === 'small' ||
+    type !== 'link' && !size ||
+    type === 'link' && size
+  ) isLimited = true;
+
+  return isLimited;
+}
+
+function getExpectedType(type, size) {
+  let expType = type;
+  if (type === 'primary' && size === 'small') { expType = ''; }
+  return expType;
+}
 
 describe('Button', () => {
-  it('does not render button without text', () => {
-    const { queryByTestId } = render(
-      <Button data-testid="no button here"></Button>
+  it('uses given text', () => {
+    muteTypeNotLinkNoSizeLog();
+    const { getByTestId } = render(
+      <Button>{TEST_TEXT}</Button>
     );
-    const el = queryByTestId('no button here');
-    expect(el).toBeNull;
+    expect(getByTestId('text').textContent).toEqual(TEST_TEXT);
   });
-  it('disables button when in loading state', () => {
-    const { queryByText } = render(
-      <Button isLoading={true}>Loading Button</Button>
-    );
-    const el = queryByText('Loading Button');
-    expect(el).toBeDisabled;
+
+  describe('icons exist as expected when', () => {
+    test('only `iconNameBefore` is given', () => {
+      muteTypeNotLinkNoSizeLog();
+      const { queryByTestId } = render(
+        <Button iconNameBefore="folder">{TEST_TEXT}</Button>
+      );
+      expect(queryByTestId('icon-before')).toBeInTheDocument();
+      expect(queryByTestId('icon-after')).not.toBeInTheDocument();
+    });
+    test('only `iconNameAfter` is given', () => {
+      muteTypeNotLinkNoSizeLog();
+      const { queryByTestId } = render(
+        <Button iconNameAfter="folder">{TEST_TEXT}</Button>
+      );
+      expect(queryByTestId('icon-before')).not.toBeInTheDocument();
+      expect(queryByTestId('icon-after')).toBeInTheDocument();
+    });
+    test('both `iconNameAfter` and `iconNameBefore` are given', () => {
+      muteTypeNotLinkNoSizeLog();
+      const { queryByTestId } = render(
+        <Button iconNameBefore="folder" iconNameAfter="file">{TEST_TEXT}</Button>
+      );
+      expect(queryByTestId('icon-before')).toBeInTheDocument();
+      expect(queryByTestId('icon-after')).toBeInTheDocument();
+    });
+  });
+
+  describe('all type & size combinations render accurately', () => {
+    it.each(BTN.TYPES)('type is "%s"', (type) => {
+      muteTypeNotLinkNoSizeLog();
+      if (isPropertyLimitation(type, TEST_SIZE)) {
+        return Promise.resolve();
+      }
```

Comment:
**Explanation:** [Postpone testing complex cases until later, in 'property limitation'](https://github.com/TACC/Core-Portal/pull/640/files#diff-9bad75f8a7092bc99d6151bfde064bf97e95ed116ccb3ed3d83b236911b90632R125-R158).